### DISCUSSION TITLE
Rewrite in rust

### DISCUSSION THREAD

**Discussion by mteam88** [pkgxdev/pkgx#366] [Ideas]
Creating this discussion to focus on some discussion in #338 about rewriting in rust.
@mxcl 
@jhheider 
@lino-levan

**Comment by mteam88** [comment]
I think that the stated 10x increase in dev time for rust vs TS (or alternative) by @jhheider here: https://github.com/teaxyz/cli/issues/338#issuecomment-1418292558
... is a good point. I, being quite bad at rust, seem to think that if I practice enough, I can improve my efficiency.

Overall dev time seems to be shorter in rust for me for very simple projects (https://softwareengineering.stackexchange.com/questions/93302/spending-too-much-time-debugging) because I spend so much less time fixing my code.

Maybe we just need to take a leap and put in the practice in rust?

**Comment by mteam88** [comment]
Also: @lino-levan mentioned:
> I am a big fan of the current plan of (nailing down the api with deno) -> (rewrite in rust) though. Seems to capture the best of both worlds.

Why (if you don't mind me asking) do we need to nail the API with deno at all? Can't some sort of language-agnostic API spec (like [Swagger](swagger.io)) do the trick? Maybe I don't understand what you mean by nailing down the API...

**Reply by lino-levan** [reply]
Ah I guess I didn't explain myself very well. Tea is a really bold new product which is making a lot of big assumptions and then testing them very quickly. Rust is fantastic when you have a project specification, you know what you want to build, and you need to build it fast and correct. Tea doesn't really fit this criteria.

Typescript (+Deno), in my opinion, is kind of the opposite of Rust in many regards. The type system is "safe enough" to ensure that a lot of obvious errors don't get through but there isn't much more safety than that. The big advantage of Typescript is that it allows for unparalleled prototyping. The code is dead simple to write, you do not have to finagle with the compiler, and in general you can write *good enough* code fast.

When Tea matures as a product, I totally understand the motivation to use Rust. Until that happens, I think that sticking with typescript and Deno makes the most sense for this project.

(Apologies if this doesn't make sense, I am writing this rather late where I am currently living)

**Reply by mteam88** [reply]
Makes sense. The prototyping aspect of TS (+Deno) is a huge advantage over rust...
